mobile/lib/components/utility/refreshable_fetch_widget.dart

```diff
@@ -0,0 +1,127 @@
+/// that pull data from an async source and need to be
+/// able to be refreshed.
+///
+/// For this widget to work, the widgets built by
+/// [builder] and [errorBuilder] _have to_ be
+/// [Scrollable], e.g. a [ListView] or a
+/// [SingleChildScrollView], so that they
+/// can be refreshed.
+class RefreshableFetchWidget<T> extends HookWidget {
+  const RefreshableFetchWidget({
+    super.key,
+    required this.defaultData,
+    required this.onFetch,
+    required this.loadingBuilder,
+    required this.errorBuilder,
+    required this.builder,
+    this.refreshIndicatorColor,
+    this.refreshIndicatorBackgroundColor,
+    this.hasData,
+  });
+
+  /// The default value of the data that is going to be fetched.
+  final T defaultData;
+
+  /// An async source of the data.
+  ///
+  /// Called once on initial render and once on every refresh.
+  final Future<T> Function() onFetch;
+
+  /// A builder that specifies the widget to be displayed while
+  /// the data is still loading.
+  final Widget Function(BuildContext, AsyncSnapshot<T>) loadingBuilder;
+
+  /// A builder that is called when the data couldn't be loaded.
+  final Widget Function(BuildContext) errorBuilder;
+
+  /// A builder that specifies a widget to be displayed on
+  /// successfully loading the data.
+  final Widget Function(BuildContext, T) builder;
+
+  /// Determines if the data was successfully loaded.
+  ///
+  /// On default, snapshot.hasData is used.
+  final bool Function(T)? hasData;
+
+  final Color? refreshIndicatorColor;
+  final Color? refreshIndicatorBackgroundColor;
+
+  @override
+  Widget build(BuildContext context) {
+    final data = _useRefreshableFetchController(defaultData, onFetch);
+
+    return RefreshIndicator(
+      color: refreshIndicatorColor ?? context.colors.highlightGreen,
+      backgroundColor:
+          refreshIndicatorBackgroundColor ?? context.colors.backgroundDark,
+      displacement: 0,
+      onRefresh: () async => data.state.value = await onFetch(),
+      child: data.snapshot.connectionState == ConnectionState.waiting
+          ? loadingBuilder(context, data.snapshot)
+          : _decideVariant(context, data),
+    );
+  }
+
+  /// Decides, which variant of child to render
+  /// based on the snapshot and the provided function
+  Widget _decideVariant(
+    BuildContext context,
+    _RefreshableFetchController<T> data,
+  ) {
+    if (hasData != null ? hasData!(data.state.value) : data.snapshot.hasData) {
+      return builder(context, data.state.value);
+    }
+    return errorBuilder(context);
+  }
+}
+
+/// Sets up hooks needed for fetching, loading and
+/// re-fetching data.
+///
+/// This function is a hook and should be used as such -
+/// at the top of the widget's `build()` function.
+///
+/// [defaultValue] - default value of the fetched data.
+///
+/// [fetchFunction] - function for fetching and re-fetching data
+_RefreshableFetchController<T> _useRefreshableFetchController<T>(
```

Comment:
haven't used hooks for a while but I think you could do this easier

Reply:
It'd be simple with 1 fetch, the reloads make it more complicated tho, that's why the `useMemoized` and `useFuture` are also used